File: hatab/src/main/resources/web/js/hatab/Horbox.js
/* Horbox.js

	Purpose:
		Horizontal Accordion Tabbox
		
	Description:
		
	History:
		Sep 16, 2010 11:35:04 AM , Created by simon

Copyright (C) 2010 Potix Corporation. All Rights Reserved.
*/
/**
 * A widget as horizontal accordion tabbox. Works like Tabbox in accordion mold,
 * but in horizontal direction. 
 * 
 * <p>Default {@link #getZclass}: z-horbox.
 */
hatab.Horbox = zk.$extends(zul.Widget, {
	
	$define: {
		/**
		 * 
		 */
		/**
		 * 
		 */
		tabWidth: _zkf = function () {
			this.rerender();
		}
		//TODO: set tab negative margin value
	},
	/**
	 * 
	 */
	getZclass: function () {
		return this._zclass == null ? "z-horbox" : this._zclass;
	},
	/**
	 * Returns the selected index.
	 * @return int
	 */
	getSelectedIndex: function() {
		return this._selPanel ? this._selPanel.getIndex() : -1 ;
	},
	/**
	 * Sets the selected index.
	 * @param int index
	 */
	setSelectedIndex: function(index) {
		this.setSelectedPanel(this.getChildAt(index));
	},
	/**
	 * Returns the selected panel.
	 * @return Panel
	 */
	getSelectedPanel: function() {
		return this._selPanel;
	},
	/**
	 * Sets the selected panel.
	 * @param Panel panel
	 */
	setSelectedPanel: function(panel) {
		if (typeof panel == 'string')
			panel = zk.Widget.$(panel);
		if (this._selPanel != panel) {
			if (panel)
				panel.setSelected(true);
				//it will set _selPanel (but we still set it later just in case)
			this._selPanel = panel;
		}
	},
	//TODO: check all below
	bind_: function (desktop, skipper, after) {
		this.$supers(hatab.Horbox, 'bind_', arguments);
		
		zWatch.listen({onResponse: this});
		
		var panel = this._selPanel;
		if (panel)
			after.push(function() {
				panel.setSelected(true);
			});
	},
	unbind_: function () {
		zWatch.unlisten({onResponse: this});
		this.$supers(hatab.Horbox, 'unbind_', arguments);
	},
	onResponse: function () {
		if (this._shallSize) {
			zWatch.fire('onSize', this);
			this._shallSize = false;
		}
	},
	_syncSize: function () {
		this._shallSize = true;
		if (!this.inServer && this.desktop)
			this.onResponse();
	},
	//super//
	removeChildHTML_: function (child) {
		this.$supers('removeChildHTML_', arguments);
		/*
		if (this.isVertical() && child.$instanceof(hatab.Hortabs))
			jq(child.uuid + '-line', zk).remove();
		*/
	},
	onChildAdded_: function (child) {
		this.$supers('onChildAdded_', arguments);
		/*
		if (child.$instanceof(hatab.Hortabs))
			this.tabs = child;
		else if (child.$instanceof(hatab.Horpanels)) {
			this.tabpanels = child;
		}
		*/
		this.rerender();
	},
	onChildRemoved_: function (child) {
		this.$supers('onChildRemoved_', arguments);
		/*
		if (child == this.toolbar)
			this.toolbar = null;
		else if (child == this.tabs)
			this.tabs = null;
		else if (child == this.tabpanels)
			this.tabpanels = null;
		if (!this.childReplacing_)
			this.rerender();
		*/
	},
	setWidth: function (width) {
		this.$supers('setWidth', arguments);
		if (this.desktop)
			zWatch.fireDown('onSize', this);
	},
	setHeight: function (height) {
		this.$supers('setHeight', arguments);
		if (this.desktop)
			zWatch.fireDown('onSize', this);
	}
});

File: hatab/src/main/java/org/zkoss/hatab/Horbox.java
/* Hatab.java

{{IS_NOTE
	Purpose:
		
	Description:
		
	History:
		Sep 16, 2010 12:08:04 PM , Created by simon
}}IS_NOTE

Copyright (C) 2010 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
}}IS_RIGHT
*/
package org.zkoss.hatab;

import java.io.IOException;

import org.zkoss.zk.ui.Component;
import org.zkoss.zk.ui.UiException;
import org.zkoss.zk.ui.sys.ContentRenderer;
import org.zkoss.zul.impl.XulElement;

/**
 * 
 * @author simon
 */
public class Horbox extends XulElement{
	
	private transient Horpanel _selPanel;
	/** The event listener used to listen onSelect for each tab. */
	///* package */transient EventListener _listener;
	
	/**
	 * 
	 * @return
	 */
	public Horpanel getSelectedPanel(){
		return _selPanel;
	}
	
	/**
	 * 
	 * @param panel
	 */
	public void setSelectedPanel(Horpanel panel){
		if (panel == null)
			throw new IllegalArgumentException("null tabpanel");
		if (panel.getHorbox() != this)
			throw new UiException("Not a child: " + panel);
		
		//_selPanel = panel;
	}
	
	/**
	 * 
	 * @return
	 */
	public int getSelectedIndex(){
		return _selPanel != null ? _selPanel.getIndex() : -1;
	}
	
	/**
	 * 
	 * @param j
	 */
	public void setSelectedIndex(int j){
		//setSelectedPanel((Horpanel) _hatabpanels.getChildren().get(j));
	}
	
	/**
	 * 
	 */
	public String getZclass() {
		return _zclass == null ? "z-horbox" : _zclass;
	}
	
	
	
	public void beforeChildAdded(Component child, Component insertBefore) {
		// TODO Auto-generated method stub
		super.beforeChildAdded(child, insertBefore);
	}
	
	public void beforeParentChanged(Component parent) {
		// TODO Auto-generated method stub
		super.beforeParentChanged(parent);
	}
	
	public boolean insertBefore(Component arg0, Component arg1) {
		// TODO Auto-generated method stub
		// select newly added tab
		return super.insertBefore(arg0, arg1);
	}

	protected void renderProperties(ContentRenderer renderer)
			throws IOException {
		super.renderProperties(renderer);
		
		// TODO
		
	}
	
}
